library/src/main/java/com/datatorrent/lib/parser/Parser.java

```diff
@@ -61,25 +70,48 @@ public void setup(PortContext context)
     }
   };
 
-  @OutputPortFieldAnnotation(optional = true)
   public transient DefaultOutputPort<INPUT> err = new DefaultOutputPort<INPUT>();
-
+  public transient DefaultOutputPort<INPUT> validatedOutput = new DefaultOutputPort<INPUT>();
   public transient DefaultInputPort<INPUT> in = new DefaultInputPort<INPUT>()
   {
     @Override
     public void process(INPUT inputTuple)
     {
-      Object tuple = convert(inputTuple);
-      if (tuple == null && err.isConnected()) {
-        err.emit(inputTuple);
-        return;
-      }
-      if (out.isConnected()) {
-        out.emit(tuple);
-      }
+      incomingTuplesCount++;
+      processTuple(inputTuple);
     }
   };
 
+  public void processTuple(INPUT inputTuple)
+  {
+    Object tuple = convert(inputTuple);
+    if (tuple == null && err.isConnected()) {
+      errorTupleCount++;
+      err.emit(inputTuple);
+      return;
+    }
+    if (out.isConnected()) {
+      emittedObjectCount++;
+      out.emit(tuple);
+    }
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    errorTupleCount = 0;
+    emittedObjectCount = 0;
+    rollingPercentValidTuples = 0;
+    rollingPercentErrorTuples = 0;
+  }
+
+  @Override
+  public void endWindow()
+  {
```

Comment:
Valid point. I added these, since it was mentioned to capture percent metrics as well. Will check what is the best way to add these. I think max and avg still will hold for percentages. Sum probably is not relevant.